nav2_costmap_2d/plugins/costmap_filters/binary_filter.cpp

```diff
@@ -0,0 +1,310 @@
+  declareParameter("binary_state_topic", rclcpp::ParameterValue("binary_state"));
+  node->get_parameter(name_ + "." + "binary_state_topic", binary_state_topic);
+
+  filter_info_topic_ = filter_info_topic;
+  // Setting new costmap filter info subscriber
+  RCLCPP_INFO(
+    logger_,
+    "BinaryFilter: Subscribing to \"%s\" topic for filter info...",
+    filter_info_topic_.c_str());
+  filter_info_sub_ = node->create_subscription<nav2_msgs::msg::CostmapFilterInfo>(
+    filter_info_topic_, rclcpp::QoS(rclcpp::KeepLast(1)).transient_local().reliable(),
+    std::bind(&BinaryFilter::filterInfoCallback, this, std::placeholders::_1));
+
+  // Get global frame required for binary state publisher
+  global_frame_ = layered_costmap_->getGlobalFrameID();
+
+  // Create new binary state publisher
+  binary_state_pub_ = node->create_publisher<nav2_msgs::msg::BinaryState>(
+    binary_state_topic, rclcpp::QoS(10));
+  binary_state_pub_->on_activate();
+
+  // Initialize state as "false" by-default
+  changeState(false);
+}
+
+void BinaryFilter::filterInfoCallback(
+  const nav2_msgs::msg::CostmapFilterInfo::SharedPtr msg)
+{
+  std::lock_guard<CostmapFilter::mutex_t> guard(*getMutex());
+
+  rclcpp_lifecycle::LifecycleNode::SharedPtr node = node_.lock();
+  if (!node) {
+    throw std::runtime_error{"Failed to lock node"};
+  }
+
+  if (!mask_sub_) {
+    RCLCPP_INFO(
+      logger_,
+      "BinaryFilter: Received filter info from %s topic.", filter_info_topic_.c_str());
+  } else {
+    RCLCPP_WARN(
+      logger_,
+      "BinaryFilter: New costmap filter info arrived from %s topic. Updating old filter info.",
+      filter_info_topic_.c_str());
+    // Resetting previous subscriber each time when new costmap filter information arrives
+    mask_sub_.reset();
+  }
+
+  if (msg->type != BINARY_FILTER) {
+    RCLCPP_ERROR(logger_, "BinaryFilter: Mode %i is not supported", msg->type);
+    return;
+  }
+
+  mask_topic_ = msg->filter_mask_topic;
+
+  // Setting new filter mask subscriber
+  RCLCPP_INFO(
+    logger_,
+    "BinaryFilter: Subscribing to \"%s\" topic for filter mask...",
+    mask_topic_.c_str());
+  mask_sub_ = node->create_subscription<nav_msgs::msg::OccupancyGrid>(
+    mask_topic_, rclcpp::QoS(rclcpp::KeepLast(1)).transient_local().reliable(),
+    std::bind(&BinaryFilter::maskCallback, this, std::placeholders::_1));
+}
+
+void BinaryFilter::maskCallback(
+  const nav_msgs::msg::OccupancyGrid::SharedPtr msg)
+{
+  std::lock_guard<CostmapFilter::mutex_t> guard(*getMutex());
+
+  if (!filter_mask_) {
+    RCLCPP_INFO(
+      logger_,
+      "BinaryFilter: Received filter mask from %s topic.", mask_topic_.c_str());
+  } else {
+    RCLCPP_WARN(
+      logger_,
+      "BinaryFilter: New filter mask arrived from %s topic. Updating old filter mask.",
+      mask_topic_.c_str());
+    filter_mask_.reset();
+  }
+
+  filter_mask_ = msg;
+  mask_frame_ = msg->header.frame_id;
+}
+
+bool BinaryFilter::transformPose(
```

Comment:
Is this function copy pasted from other filters? We could add this to the filter base class. Seems generally useful.

Reply:
Okay, done in https://github.com/ros-planning/navigation2/pull/3228/commits/9fba2a9bff21d33455daf83e24142179552d03ec